### PR TITLE
Add zero copy mechanism for inserting entries into trace buffer

### PR DESCRIPTION
**Pull request by mmkayPL**
Signed-off-by: Mateusz Kozlowski <mateusz.kozlowski@intel.com>

### CODE REVIEW

File: source/octf/trace/trace.h [hunk 1]
```diff
@@ -138,6 +140,39 @@ int64_t octf_trace_get_lost_count(octf_trace_t trace);
...
+ *
+ * @note Only producer may mark the event.
+ *
+ * @param[in] trace Trace handle
+ * @param[in] event_t Handle to event
+ *
+ * @retval 0 Event commited successfully
+ * @retval -EINVAL Trace or handle is invalid
+ */
+int octf_trace_commit_wr_buffer(octf_trace_t trace, octf_trace_event_handle_t event_t);
```
Comment (mariuszbarczak) [inline source/octf/trace/trace.h]: ev_hndl

File: tests/octf/trace/TracingTest.cpp [hunk 2]
```diff
@@ -402,6 +524,52 @@ class Worker {
...
+            result = octf_trace_commit_wr_buffer(m_log, ev_t);
+
+            if (0 == result) {
+                env_atomic64_inc(&m_eventCount);
+            }
+        }
+    }
+
+private:
+    bool isEventLimitReached() const {
```
Comment (mariuszbarczak) [inline tests/octf/trace/TracingTest.cpp]: Can you check if this function can be reused form base class

File: tests/octf/trace/TracingTest.cpp [hunk 3]
```diff
@@ -402,6 +524,52 @@ class Worker {
...
+            : Worker(log, eventLog) {}
+
+    virtual ~EventCopier() {}
+
+private:
+    void work() {
+        size_t minEventSize = Event::getMinEventSize();
+        while (isTestRunning() && !isEventLimitReached()) {
+            octf_trace_event_handle_t ev_t;
+            uint32_t size =
```
Comment (mariuszbarczak) [inline tests/octf/trace/TracingTest.cpp]: std::max

File: tests/octf/trace/TracingTest.cpp [hunk 4]
```diff
@@ -67,11 +67,16 @@ class Event {
         setSeqId();
     }
 
-    Event(void *buffer, uint32_t size)
+    Event(void *buffer, uint32_t size, bool genId = false)
```
Comment (mariuszbarczak) [inline tests/octf/trace/TracingTest.cpp]: Check if size meet min

File: tests/octf/trace/TracingTest.cpp [hunk 5]
```diff
@@ -67,11 +67,16 @@ class Event {
...
             : m_buffer(nullptr)
             , m_size(size) {
         m_buffer = new char[m_size];
         memcpy_s(m_buffer, m_size, buffer, size);
+        if (genId) {
```
Comment (mariuszbarczak) [inline tests/octf/trace/TracingTest.cpp]: I propose to introduce a private utils method which initialize buffer.

File: tests/octf/trace/TracingTest.cpp (hunk 4 above)
Comment (mariuszbarczak) [inline tests/octf/trace/TracingTest.cpp]: let's avoid default value for argument

File: tests/octf/trace/TracingTest.cpp [hunk 6]
```diff
@@ -48,29 +53,31 @@ class Event {
...
-        setSeqId();
+        initializeBuffer(m_buffer, m_size);
     }
 
     Event(uint32_t size)
             : m_buffer(nullptr)
             , m_size(size) {
         if (m_size < sizeof(EventSeqId)) {
-            m_size = sizeof(EventSeqId);
+            throw octf::Exception("Invalid event size");
```
Comment (mariuszbarczak) [inline tests/octf/trace/TracingTest.cpp]: Can we assert with GTEST macro?

File: tests/octf/trace/TracingTest.cpp [hunk 7]
```diff
@@ -48,29 +53,31 @@ class Event {
...
-        setSeqId();
+        initializeBuffer(m_buffer, m_size);
     }
 
-    Event(void *buffer, uint32_t size)
+    Event(void *buffer, uint32_t size, bool genId)
             : m_buffer(nullptr)
             , m_size(size) {
+        if (m_size < sizeof(EventSeqId)) {
+            throw octf::Exception("Invalid event size");
```
Comment (mariuszbarczak) [inline tests/octf/trace/TracingTest.cpp]: Can we assert, fail with GTEST macro?

File: source/octf/trace/trace.c [hunk 8]
```diff
@@ -421,6 +421,66 @@ int octf_trace_push(octf_trace_t trace,
     return 0;
 }
 
+int octf_trace_get_wr_buffer(octf_trace_t trace,
+                             octf_trace_event_handle_t *ev_hndl,
```
Comment (mariuszbarczak) [inline source/octf/trace/trace.c]: Maybe in case of error we can zero output arguments?

File: tests/octf/trace/TracingTest.cpp (hunk 7 above)
Comment (mmkayPL) [reply tests/octf/trace/TracingTest.cpp]: Assert and fail return a value, so can't use them in the constructor

File: tests/octf/trace/TracingTest.cpp (hunk 6 above)
Comment (mmkayPL) [reply tests/octf/trace/TracingTest.cpp]: Assert and fail return a value, so can't use them in the constructor

**Pull request merged**